SI-7290 Discard duplicates in switchable alternative patterns.

retronym opened the pull request:
The pattern matcher must not allow duplicates to hit the
backend when generating switches. It already eliminates then
if they appear on different cases (with an unreachability warning.)

This commit does the same for duplicated literal patterns in an
alternative pattern: discard and warn.

Review by @adriaanm

retronym committed:
SI-7290 Discard duplicates in switchable alternative patterns.

The pattern matcher must not allow duplicates to hit the
backend when generating switches. It already eliminates then
if they appear on different cases (with an unreachability warning.)

This commit does the same for duplicated literal patterns in an
alternative pattern: discard and warn.

paulp commented on src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala:
```diff
@@ -22,7 +22,7 @@ import scala.reflect.internal.util.NoPosition
   import PatternMatchingStats._
   import global.{Tree, Type, Symbol, NoSymbol, CaseDef, atPos,
-    ConstantType, Literal, Constant, gen, EmptyTree,
+    ConstantType, Literal, Constant, gen, EmptyTree, distinctBy,
```
NICE USE OF distinctBy! You Might Be The First!

adriaanm commented on src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala:
```diff
@@ -452,7 +452,17 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
+                    case _              => t
+                  }
+                  // SI-7290 Discard duplicate alternatives that would crash the backend
+                  def distinctAlts = distinctBy(switchableAlts)(extractConst)
```
why is this a def? always evaluated twice

adriaanm commented on src/compiler/scala/tools/nsc/transform/patmat/MatchOptimization.scala:
```diff
@@ -452,7 +452,17 @@ trait MatchOptimization extends MatchTreeMaking with MatchAnalysis {
+                  def distinctAlts = distinctBy(switchableAlts)(extractConst)
+                  if (distinctAlts.size < switchableAlts.size) {
+                    val duplicated = switchableAlts.groupBy(extractConst).flatMap(_._2.drop(1))
+                    global.currentUnit.warning(pos, s"Pattern contains duplicate alternatives: ${duplicated.mkString(", ")}")
```
I'd only report distinct duplicates, if you know what I mean :neckbeard:

retronym committed:
SI-7290 Minor cleanups driven by review comments.

  - make a def a val, we only need to compute it once
  - add a clarifying comment
  - only report the first duplicate

retronym commented:
@adriaanm done
